### PR TITLE
Fix: Duplicate Success Message Printing (Entity Generation) #11502

### LINKED ISSUE #11502: Duplicate Success Message Printing: Entity Generation
##### **Overview of the issue**

When generating entities using a JDL file, upon success we print the message; `INFO! Congratulations, JHipster execution is complete!`. However this message seems to be printed for each entity generated. So for example if I have three entities in my JDL file that message is printed three times. 

##### **Motivation for or Use Case**

It's not the most visually pleasing when say we have a JDL file with 10 entities and the success message being printed 10 times. 

##### **Reproduce the error**

Create a simple project. You can use the below configuration. 

##### **Related issues**

None.

##### **Suggest a Fix**

Don't have one right now. If someone is willing to take this please feel free to do so; otherwise I'll fix this in the future. :smile: 

##### **JHipster configuration**

<pre>&lt;details&gt;
&lt;summary&gt;.yo-rc.json file&lt;/summary&gt;
&lt;pre&gt;
{
  &quot;generator-jhipster&quot;: {
    &quot;promptValues&quot;: {
      &quot;packageName&quot;: &quot;com.mycompany.myapp&quot;
    },
    &quot;jhipsterVersion&quot;: &quot;6.8.0&quot;,
    &quot;applicationType&quot;: &quot;monolith&quot;,
    &quot;baseName&quot;: &quot;jhipster&quot;,
    &quot;packageName&quot;: &quot;com.mycompany.myapp&quot;,
    &quot;packageFolder&quot;: &quot;com/mycompany/myapp&quot;,
    &quot;serverPort&quot;: &quot;8080&quot;,
    &quot;authenticationType&quot;: &quot;jwt&quot;,
    &quot;cacheProvider&quot;: &quot;ehcache&quot;,
    &quot;enableHibernateCache&quot;: false,
    &quot;websocket&quot;: false,
    &quot;databaseType&quot;: &quot;sql&quot;,
    &quot;devDatabaseType&quot;: &quot;h2Disk&quot;,
    &quot;prodDatabaseType&quot;: &quot;mysql&quot;,
    &quot;searchEngine&quot;: false,
    &quot;messageBroker&quot;: false,
    &quot;serviceDiscoveryType&quot;: false,
    &quot;buildTool&quot;: &quot;maven&quot;,
    &quot;enableSwaggerCodegen&quot;: false,
    &quot;jwtSecretKey&quot;: &quot;bXktc2VjcmV0LXRva2VuLXRvLWNoYW5nZS1pbi1wcm9kdWN0aW9uLWFuZC10by1rZWVwLWluLWEtc2VjdXJlLXBsYWNl&quot;,
    &quot;embeddableLaunchScript&quot;: false,
    &quot;useSass&quot;: true,
    &quot;clientPackageManager&quot;: &quot;npm&quot;,
    &quot;clientFramework&quot;: &quot;angularX&quot;,
    &quot;clientTheme&quot;: &quot;none&quot;,
    &quot;clientThemeVariant&quot;: &quot;&quot;,
    &quot;creationTimestamp&quot;: 1585028296449,
    &quot;testFrameworks&quot;: [],
    &quot;jhiPrefix&quot;: &quot;jhi&quot;,
    &quot;entitySuffix&quot;: &quot;&quot;,
    &quot;dtoSuffix&quot;: &quot;DTO&quot;,
    &quot;otherModules&quot;: [],
    &quot;enableTranslation&quot;: false,
    &quot;blueprints&quot;: []
  }
}

&lt;/pre&gt;
&lt;/details&gt;

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

&lt;details&gt;
&lt;summary&gt;JDL entity definitions&lt;/summary&gt;

&lt;pre&gt;
entity Region {
  regionName String
}
entity Country {
  countryName String
}
entity Location {
  streetAddress String,
  postalCode String,
  city String,
  stateProvince String
}

&lt;/pre&gt;
&lt;/details&gt;

##### **Environment and Tools**

openjdk version &quot;11.0.6&quot; 2020-01-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.6+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.6+10, mixed mode)

git version 2.17.1

node: v12.13.0

npm: 6.12.0

yarn: 1.21.1

Docker version 19.03.8, build afacb8b7f0

docker-compose version 1.24.1, build 4667896b
</pre>

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by qmonmert**
Fix #11502

-   [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: cli/import-jdl.js [hunk 1]
```diff
@@ -229,14 +229,15 @@ const generateEntityFiles = (generator, entity, inFolder, env, shouldTriggerInst
...
+        const callback = isLastEntityToGenerate ? done : () => {};
         env.run(
             command,
             {
                 ...options,
                 force: options.force || !options.interactive,
                 'skip-install': !shouldTriggerInstall
             },
-            done
+            callback
```
Comment (SudharakaP) [inline cli/import-jdl.js]: This is minor but the maybe we can simplify this by just removing the constant and callback and adding; 

```js
if (isLastEntityToGenerate) 
    done
```

**Comment by mshima** [comment]
I didn't test, but I think this PR still doesn't fix the duplicate problem.

The last entity generation will print the message.
Then the jhipster client/server workflow will continue and print the message again.

**Comment by SudharakaP** [comment]
@mshima : Good point. Let me manually test it to verify. :smile:

File: cli/import-jdl.js (hunk 1 above)
Comment (qmonmert) [reply cli/import-jdl.js]: @SudharakaP do you mean ? 
`done`
=>
`isLastEntityToGenerate ? done : () => {}`

**Comment by qmonmert** [comment]
@mshima & @SudharakaP I tested but you can check

File: cli/import-jdl.js (hunk 1 above)
Comment (qmonmert) [reply cli/import-jdl.js]: because we can't write your code like this

**Comment by mshima** [comment]
@qmonmert it will ignore the error message if it exists.

**Comment by mshima** [comment]
Here is my proposed solution:
https://github.com/mshima/generator-jhipster/commit/8e1a2d465c1da6002e9966499978b78fad0d90f6

File: cli/import-jdl.js (hunk 1 above)
Comment (SudharakaP) [reply cli/import-jdl.js]: @qmonmert : Got it; you are correct. I didn't see it's an argument. :smile:

**Comment by SudharakaP** [comment]
I've tested it and it fixes the error messages; but the error handling that @mshima mentioned is a concern I guess.

**Comment by mshima** [comment]
Others problem we should take in account:
- jdl with multiples application.
- jdl with deployment.

@MathieuAA jdl supports combination of multiples applications, multiples deployments or this kind of use cases?

**Comment by MathieuAA** [comment]
Yup

**Comment by qmonmert** [comment]
@mshima & @SudharakaP I updated my PR to display errors if import fails but do you need to do "jdl supports combination of multiples applications, multiples deployments or this kind of use cases" in this PR ? If yes, what is the scenario ? cc @MathieuAA

**Comment by mshima** [comment]
@qmonmert actually it’s done:
https://github.com/jhipster/generator-jhipster/pull/11509

JDL supports multiple combination of applications, entities and deployments.
To get a success message I had to migrate to promises.

**Comment by mshima** [comment]
Your PR works for entities with 1 application which the bug report is about.

**Comment by SudharakaP** [comment]
@qmonmert : Just to add to what mshima said; here's an example of a JDL which includes more than one application; https://www.jhipster.tech/jdl/applications#more-than-one-application

Maybe @mshima can confirm, but I think unfortunately (:cry: :cry:) this PR should be closed once #11509 is merged as it covers both cases.

**Comment by qmonmert** [comment]
@SudharakaP ok to close this PR 😢

**Pull request closed without merging**